llvm/lib/Transforms/Scalar/Scalarizer.cpp

```diff
@@ -699,6 +730,22 @@ bool ScalarizerVisitor::splitCall(CallInst &CI) {
   if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))
     Tys.push_back(VS->SplitTy);
 
+  if (AreAllMatchingVectors) {
+    Type *PrevType = CallType->getContainedType(0);
+    for (unsigned I = 1; I < CallType->getNumContainedTypes(); I++) {
+      Type *CurrType = cast<FixedVectorType>(CallType->getContainedType(I));
+      if (PrevType != CurrType) {
```

Comment:
So, this enforces the same component type across all returned vector results, which answers the question in the prior comment: "Should they match the overload type, or can they be different?".

However, I'm not sure it's necessary for them all to have matching component types, is it?  Though if they differ in bit width, this could create a problem when using `ScalarizeMinBits` to prevent scalarization of small types, since this could create a different `NumPacked` for different returned results.

This same issue would apply to input arguments, and now that I look at how that's handled, I see the comment that it won't handle it if `NumPacked` for the op is different than for `VS` (returned result).

I think it would be ok to allow different components, as long as incorporating the same check as is used for the operands instead of expecting an exact match in type.

Reply:
I would prefer to be more conservative. For one I don't think its necessary to relax this because any different component types should have been determined on the vectorized form of the intrinsic. Second, we don't need anything but an exact component match for our use case so why increase our test burden when it isn't necessary?

Reply:
Wait, I was confused thinking this was comparing equal.  So, `frexp` returns two different vector types.  The prior call to `isVectorIntrinsicWithOverloadTypeAtArg` determines whether the first type is part of the overload, but then we assume any other types in the return that don't match the first one are also part of the overload type.  That doesn't sound right for the general case either.

Since we don't actually ensure that they match, we could have a situation where `NumPacked` differs between returned vector types, which would be a problem.  This was checked for arguments, bailing if they differ, but we aren't checking this for the return values.  I think we need to do that as well.

Reply:
I think you are still confused. `isVectorIntrinsicWithOverloadTypeAtArg ` doesn't check overloads for return types. Go look at that code path return will always be true.

Reply:
Will discuss directly.